src/fsspec_xrootd/xrootd.py

```diff
@@ -219,6 +321,163 @@ def ls(self, path: str, detail: bool = True, **kwargs: Any) -> list[Any]:
+                raise OSError(f"File failed to read: {status.message}")
+            status, data = await _async_wrap(
+                _myFile.read,
+                start,
+                end - start,
+                self.timeout,
+            )
+            if not status.ok:
+                raise OSError(f"Bytes failed to read from open file: {status.message}")
+            return data
+        except Exception as exc:
+            status, _n = await _async_wrap(
+                _myFile.close,
+                self.timeout,
+            )
+            raise OSError(f"Something went wrong in _cat_file(): {exc}")
+
+    async def _get_max_chunk_info(self, file: Any) -> tuple[int, int]:
+        data_server = file.get_property("DataServer")
+        if data_server not in XRootDFileSystem.cat_chunk_cache.keys():
+            status, result = await _async_wrap(
+                self._myclient.query, QueryCode.CONFIG, "readv_iov_max readv_ior_max"
+            )
+            if not status.ok:
+                raise OSError(
+                    f"Server query for vector read info failed: {status.message}"
+                )
+            max_num_chunks, max_chunk_size = map(int, result.split(b"\n", 1))
+            XRootDFileSystem.cat_chunk_cache.update(
+                {
+                    data_server: {
+                        "max_num_chunks": max_num_chunks,
+                        "max_chunk_size": max_chunk_size,
+                    }
+                }
+            )
+        return (
+            int(XRootDFileSystem.cat_chunk_cache[data_server]["max_num_chunks"]),
+            int(XRootDFileSystem.cat_chunk_cache[data_server]["max_chunk_size"]),
+        )
+
+    async def _cat_vector_read(
+        self,
+        path: str,
+        chunks: list[tuple[int, int]],
+        batch_partition: int,
+    ) -> dict[str, list[bytes]]:
+        try:
+            _myFile = client.File()
+            status, _n = await _async_wrap(
+                _myFile.open,
+                self.protocol + "://" + self.storage_options["hostid"] + "/" + path,
+                OpenFlags.READ,
+                self.timeout,
+            )
+            if not status.ok:
+                raise OSError(f"File did not open properly: {status.message}")
+
+            max_num_chunks, max_chunk_size = await self._get_max_chunk_info(_myFile)
+            vectors = _make_vectors(chunks, max_num_chunks, max_chunk_size)
+
+            coros = [_async_wrap(_myFile.vector_read, v, self.timeout) for v in vectors]
+
+            results = await _run_coros_in_chunks(
+                coros, batch_size=batch_partition, nofiles=True
+            )
+
+            deets = []
+            for result in results:
+                if not result[0].ok:
+                    raise OSError(
+                        f"File did not vector_read properly: {status.message}"
+                    )
+                for r in result[1]:
+                    deets.append(r.buffer)
```

Comment:
I would re-assemble the originally requested chunks here, rather than in the parent call. e.g.
```python
deets = [b""]
ichunk = 0
for status, buffers in results:
    if not status.ok:
        ...
    for buf in buffers:
        if buf.offset < chunks[ichunk][1]:
            deets[-1] += buf.buffer
        else:
            deets.append(buf.buffer)
            ichunk += 1
```

Reply:
this will work overall, but I think you need to change .offset to .length. Also chunks is in the wrong format (start, end) while it seems you are expecting (start, length). Instead of using chunks, we should use each vector in the vectors list. I will make the changes and show you in the next version of this PR

Reply:
The point here is we need to figure out when we are appending to a previous chunk and when we are starting a new one. So I check if the current offset of the returned buffer is within the range of the current chunk, and if not start a new one (under the assumption that split chunks are at least returned in order).